### PR TITLE
cdc: incremental scans acquire snapshots before semaphores to avoid useless queue

### LINKED ISSUE #15866: cdc incremental scan tasks can't be canceld by region split or merge correctly
### What version of TiKV are you using?
Master.

### What operating system and CPU are you using?
Doesn't matter.

### Steps to reproduce
Send lots of cdc incremental scan tasks to a TiKV instance, which has lots of region split or merge events.

### What did you expect?

Incremental scan tasks can be canceld by region split or merge quickly in all situations.

### What did happened?

Some incremental scan tasks can't be canceld quickly, because they are blocking on `concurrency_semaphore_.acquire().await`.

### PR DESCRIPTION
**Pull request by hicqu**
### What is changed and how it works?

Issue Number: Close #15866 

What's Changed:

Without the PR incremental scans acquire snapshots after get tokens. So if there are lots of pending incremental scan tasks, they will queue before get the token even if the region splits or merges.

This PR makes incremental scans acquire snapshots before get tokens. Then if region splits or merges, the task can know the event and cancel the task directly.

```commit-message
 cdc: incremental scans acquire snapshots before semaphores to avoid useless queue
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/initializer.rs [hunk 1]
```diff
@@ -107,30 +108,9 @@ impl<E: KvEngine> Initializer<E> {
...
     ) -> Result<()> {
         fail_point!("cdc_before_initialize");
-        let _permit = concurrency_semaphore.acquire().await;
-
-        // When downstream_state is Stopped, it means the corresponding delegate
-        // is stopped. The initialization can be safely canceled.
-        //
-        // Acquiring a permit may take some time, it is possible that
-        // initialization can be canceled.
-        if self.downstream_state.load() == DownstreamState::Stopped {
```
Comment (3pointer) [inline components/cdc/src/initializer.rs:121]: shall we also keep the check here? just remove the _permit.

**Review by 3pointer — COMMENTED** [review]
rest LGTM

**Review by 3pointer — APPROVED** [review]

File: components/cdc/src/initializer.rs [hunk 2]
```diff
@@ -186,6 +166,11 @@ impl<E: KvEngine> Initializer<E> {
     ) -> Result<()> {
         if let Some(region_snapshot) = resp.snapshot {
             assert_eq!(self.region_id, region_snapshot.get_region().get_id());
+            CDC_SCAN_TASKS.with_label_values(&["ongoing"]).inc();
```
Comment (pingyu) [inline components/cdc/src/initializer.rs:169]: How about moving these 4 lines to the `async_incremental_scan`? That would enhance the cohesion of the `async_incremental_scan` function.

**Review by pingyu — COMMENTED** [review]
Rest LGTM~

File: components/cdc/src/initializer.rs (hunk 1 above)
Comment (hicqu) [reply components/cdc/src/initializer.rs:121]: It will be checked later soon. So we can remove this pre-check to simplify code.

**Review by pingyu — APPROVED** [review]

**Pull request merged**